knowhere/index/vector_index/IndexDiskANN.cpp

```diff
@@ -0,0 +1,269 @@
+// Copyright (C) 2019-2020 Zilliz. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance
+// with the License. You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software distributed under the License
+// is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+// or implied. See the License for the specific language governing permissions and limitations under the License.
+
+#include "index/vector_index/IndexDiskANN.h"
+
+#include <omp.h>
+
+#include <limits>
+#include <sstream>
+#include <utility>
+#include <vector>
+
+#include "DiskANN/include/aux_utils.h"
+#ifndef _WINDOWS
+#include "DiskANN/include/linux_aligned_file_reader.h"
+#else
+#include "DiskANN/include/windows_aligned_file_reader.h"
+#endif
+#include "knowhere/common/Exception.h"
+#include "knowhere/index/vector_index/IndexDiskANNConfig.h"
+#include "knowhere/index/vector_index/adapter/VectorAdapter.h"
+
+namespace knowhere {
+
+template <typename T>
+IndexDiskANN<T>::IndexDiskANN(std::string index_prefix, MetricType metric_type,
+                              std::unique_ptr<FileManager> file_manager)
+    : index_prefix_(index_prefix), file_manager_(std::move(file_manager)) {
+    index_type_ = IndexEnum::INDEX_DISKANN;
+
+    if (metric_type == metric::L2) {
+        metric_ = diskann::L2;
+    } else if (metric_type == metric::IP) {
+        if (!std::is_same_v<T, float>) {
+            KNOWHERE_THROW_MSG(
+                "DiskANN currently only supports floating point data for Max "
+                "Inner Product Search. ");
+        }
+        metric_ = diskann::INNER_PRODUCT;
+    } else {
+        KNOWHERE_THROW_MSG("DiskANN only support L2 and IP distance.");
+    }
+}
+
+namespace {
+void
+CheckPreparation(bool is_prepared) {
+    if (is_prepared) {
```

Comment:
is this logic correct ?

Reply:
good catch! I am working on the UT to cover these noob problems